diff --git a/regenpfeifer/stroke_generator.py b/regenpfeifer/stroke_generator.py
index e35c6ba..e84630c 100644
--- a/regenpfeifer/stroke_generator.py
+++ b/regenpfeifer/stroke_generator.py
@@ -39,7 +39,7 @@ class StrokeGenerator(object):
         word = word.lower()
         
         splitted_word = self.word_splitter.split(word)
-        aggregated_words = self.stroke_aggregator.aggregate_strokes('/'.join(splitted_word))
+        aggregated_words = ['/'.join(splitted_word)] #self.stroke_aggregator.aggregate_strokes('/'.join(splitted_word))
         
         matched_strokes_list = []
         # for element in itertools.product(*somelists):
